modules/core/src/main/scala/zio/kafka/streams/TopologySdk.scala

```diff
@@ -0,0 +1,69 @@
+package zio.kafka.streams
+
+import org.apache.kafka.common.serialization.Serde
+import org.apache.kafka.streams.Topology
+import org.apache.kafka.streams.scala.StreamsBuilder
+import org.apache.kafka.streams.scala.kstream.{ Consumed, KStream, Produced }
+import zio.prelude.State
+import zio.prelude.fx.ZPure
+
+trait TopologySdk {
+  def streamFromSource[K, V](
+    sourceSafeTopic: SafeTopic[K, V]
+  ): State[StreamsBuilder, KStream[K, V]] = {
+    val consumed = Consumed.`with`(sourceSafeTopic.keySerde, sourceSafeTopic.valueSerde)
+    ZPure
+      .get[StreamsBuilder]
+      .map(_.stream[K, V](sourceSafeTopic.topicName)(consumed))
+  }
+
+  def streamSinkTo[K, V](
+    stream: KStream[K, V],
+    sinkSafeTopic: SafeTopic[K, V]
+  ): ZPure[StreamsBuilder, StreamsBuilder, Any, Nothing, Unit] = {
+    val produced = Produced.`with`(sinkSafeTopic.keySerde, sinkSafeTopic.valueSerde)
+    ZPure
+      .unit[StreamsBuilder]
+      .as(stream.to(sinkSafeTopic.topicName)(produced))
+  }
+
+  def branch[K, V](stream: KStream[K, V])(
+    predicate1: (K, V) => Boolean,
+    predicate2: (K, V) => Boolean
+  ): State[StreamsBuilder, Map[(K, V) => Boolean, KStream[K, V]]] =
+    ZPure.unit[StreamsBuilder].as {
+      val branches = stream.branch(predicate1, predicate2)
+      Map(
+        predicate1 -> branches(0),
+        predicate2 -> branches(1)
+      )
+    }
+
+  def safeBranch[K, Vin, Vout1 <: Vin, Vout2 <: Vin](stream: KStream[K, Vin])(
+    extractor1: Extractor[K, Vin, Vout1],
+    extractor2: Extractor[K, Vin, Vout2]
+  ): State[StreamsBuilder, (KStream[K, Vout1], KStream[K, Vout2])] =
+    ZPure.unit[StreamsBuilder].as {
+      val branches = stream.branch(extractor1.predicate, extractor2.predicate)
+      val stream1  = branches(0).mapValues(extractor1.downcaster)
```

Comment:
If you describe the topology you should expect only a branch node, in this case, you would see 2 more. What's the cost of a `mapValues` to be able to "cast" a type?

Reply:
I thought about this. I think that the cost of mapping and create a `branchMap` is not the issue here. We can always give users the ability to use plain `branch`.

The real issue here is that it's not even necessary to map inside this branch. In theory we should be able to use the `downcaster` function **without** kafka streams `mapValues` but just as a regular function. Moreover it would be deleted by the compiler because of type erasure.

It will be a bit hacky maybe, but it's guaranteed to work and remove the cost of an additional 2 nodes.